deltacat/utils/metrics.py

```diff
@@ -0,0 +1,108 @@
+    return metrics_name_with_type
+
+
+def _get_current_node_ip() -> str:
+    current_node_ip = urllib.request.urlopen(
+        "http://169.254.169.254/latest/meta-data/local-ipv4"
+    ).read().decode("utf-8")
+    return {"Name": f"node_ip",
+            "Value": f"{current_node_ip}"
+            }
+
+
+def _get_current_task_id() -> Dict[str, Any]:
+    return {"Name": f"ray_task_id",
+            "Value": f"{ray.get_runtime_context().get_task_id()}"
+            }
+
+
+def _get_current_worker_id() -> Dict[str, Any]:
+    return {"Name": f"ray_worker_id",
+            "Value": f"{ray.get_runtime_context().worker.core_worker.get_worker_id()}"
+            }
+
+
+def _get_current_instance_region() -> str:
+    current_instance_region = urllib.request.urlopen(
+        "http://169.254.169.254/latest/meta-data/placement/region"
+    ).read().decode("utf-8")
+    return current_instance_region
+
+
+def _get_current_cluster_name() -> str:
+    with open("/home/ubuntu/ray_bootstrap_config.yaml", "r") as f:
+        cluster_config = yaml.load(f)
+    return cluster_config["cluster_name"]
+
+
+METRICS_DIMENSION_TYPE_TO_VALUE_DICT: Dict[
+            str, Callable] = {
+    MetricsDimensionType.NODE_IP.value: _get_current_node_ip,
+    MetricsDimensionType.RAY_TASK_ID.value: _get_current_task_id,
+    MetricsDimensionType.RAY_WORKER_ID.value: _get_current_worker_id,
+}
+
+
+def _build_cloudwatch_metrics(metrics_name: str, metrics_type: Enum, value: str, dimension_types: List[Enum], timestamp, **kwargs) -> Dict[str, Any]:
+    metrics_name_with_type = _build_metrics_name(metrics_type, metrics_name)
+    dimensions = []
+    for dimension_type in dimension_types:
+        dimensions.append(METRICS_DIMENSION_TYPE_TO_VALUE_DICT.get(dimension_type.value)())
+    return [
+        {
+            "MetricName": f"{metrics_name_with_type}",
+            "Dimensions": dimensions,
+            "Timestamp": timestamp,
+            "Value": value,
+            **kwargs
+        }
+    ]
+
+
+def emit_metrics(metrics_name: str, metrics_type: Enum, value: str, dimension_types: List[Enum], timestamp, **kwargs) -> None:
+    current_instance_region = _get_current_instance_region()
+    cloudwatch_resource = resource_cache("cloudwatch", current_instance_region)
+    cloudwatch_client = cloudwatch_resource.meta.client
+    metrics_data = _build_cloudwatch_metrics(metrics_name,
+                                             metrics_type,
+                                             value,
+                                             dimension_types,
+                                             timestamp,
+                                             **kwargs)
+    cluster_name = _get_current_cluster_name()
+    try:
+        response = cloudwatch_client.put_metric_data(
+            Namespace=f"DEFAULT_DELTACAT_METRICS_NAMESPACE_{cluster_name}",
+            MetricData=metrics_data)
+    except ClientError as e:
```

Comment:
Not only `ClientError`, do we want to fail compaction if emitting metrics fails? I guess not.

Reply:
Yeah, should not fail the whole compaction workflow. Will catch other all exceptions here.

Reply:
correct me if I'm wrong. Cluster name might be same across jobs if we want to speed up cluster provision with cached instances.

Reply:
Yeah, if going forward we need to use cached instances, the cluster name will not be a good identifier here. We used to design the cluster name as an unique mapping to each job run. 
At the end of the day, the orchestrator launches the job should know the unique identifier for us to detect which cluster and job instructions this cluster is based on. So how about I add a variable here as `job_run_id` and let orchestrator fill in this environment variable?